Map and debug PlutusFailure cases of ApplyTx

ch1bo opened the pull request:
When submitting transactions which spend from scripts to the L2 cardano ledger, we can use debugPlutus to re-evaluate the script and dump debug information.

This does not fix the formatting though (newlines in Text).

TODO: a test would be good. @noonio you had a e2e scenario where we spend a script on the L2 recently?

This commit is also back-ported onto `doom` to benefit the hydra-doom project.

---

* [x] CHANGELOG updated
* [ ] Documentation updated or not needed
* [x] Haddocks update not needed
* [x] No new TODOs introduced

ch1bo committed:
Map and debug PlutusFailure cases of ApplyTx

When submitting transactions which spend from scripts to the L2 cardano
ledger, we can use debugPlutus to re-evaluate the script and dump debug
information.

This does not fix the formatting though (newlines in Text).

noonio reviewed (approved):
Looks great!

ch1bo commented:
> Looks great!

@noonio think we should merge without a test case? Was manually confirmed working by hydra doom people, but without a test it's (too) easy to remove when ledger types will change and annoy (definitely going to happen).

noonio commented:
> > Looks great!
> 
> @noonio think we should merge without a test case? Was manually confirmed working by hydra doom people, but without a test it's (too) easy to remove when ledger types will change and annoy (definitely going to happen).

We can't merge the test-case I have; but I'm happy to rebase and see how it plays out with that one once it's in.

I'm happy to merge as-is; (when it compiles 😅).

If you have in mind a much simpler test-case than an end-to-end one I'm very open to that!

v0d1ch commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -68,7 +79,16 @@ cardanoLedger globals ledgerEnv =
+          "Plutus validation failed: "
+            <> msg
+            <> "Debug info: "
+            <> show (debugPlutus @StandardCrypto (decodeUtf8 ctx))
```
`debugPlutus` returns IO action, perhaps things have changed in the meantime and the signature changed and now it seems we need `unsafePerformIO` or similar to display plutus falilure.

ch1bo replied:
Yeah... I might ask the cardano-ledger maintainers by sending a PR.

ch1bo replied:
Asked on discord: https://discord.com/channels/1136727663583698984/1239889030984896614/1330985414680510556

ch1bo committed:
Use unsafeDupablePerformIO to debugPlutus

It's not clear why debugPlutus would need to be an IO action, it does
not have any side effects.

See also https://github.com/IntersectMBO/cardano-ledger/commit/8b5fa4a26a69e4089ca1dc61e73d09116a9c4edc
